### ISSUE TITLE
reg_resample hang when resampling tensors

### ISSUE THREAD

**Issue by mmodat** [KCL-BMEIS/niftyreg#14]
In gitlab by @imalone on Feb 15, 2017, 13:51

@mmodat  I've found that reg_resample hangs when it encounters a non-invertible tensor as Eigen's assumptions for .log aren't satisfied and it can't converge. Adding a check on the determinant fixes the cases I've run across so far.

```
index e0802d1..3c87881 100644
--- a/reg-lib/cpu/_reg_maths_eigen.cpp
+++ b/reg-lib/cpu/_reg_maths_eigen.cpp
@@ -282,6 +282,7 @@ void reg_mat33_logm(mat33 *in_tensor)
 
    // Convert to Eigen format
    bool all_zeros = true;
+   double det = 0;
    for (sm = 0; sm < 3; sm++){
       for (sn = 0; sn < 3; sn++){
          float val=in_tensor->m[sm][sn];
@@ -290,7 +291,14 @@ void reg_mat33_logm(mat33 *in_tensor)
          tensor(sm, sn) = static_cast<double>(val);
       }
    }
-   if(all_zeros==true){
+   // Actually R case requires invertible and no negative real ev,
+   // but the only observed case so far was non-invertible.
+   // determinant is not a perfect check for invertibility and
+   // identity with zero not great either, but the alternative
+   // is a general eigensolver and the logarithm function should
+   // succeed unless convergence just isn't happening.
+   det = tensor.determinant();
+   if(all_zeros==true || det == 0){
       reg_mat33_to_nan(in_tensor);
       return;
    }
```

**Comment by mmodat** [comment]
In gitlab by @mmodat on Feb 15, 2017, 15:01

You should be able to create a branch a do a merge request. Just make sure it stays Windows/Clang compatible :)

Some things for the command display

**Comment by mmodat** [comment]
In gitlab by @imalone on Feb 15, 2017, 17:10

mentioned in merge request !9

**Comment by mmodat** [comment]
In gitlab by @mmodat on Feb 16, 2017, 17:42

closed via merge request !9

**Comment by mmodat** [comment]
In gitlab by @mmodat on Feb 16, 2017, 17:42

closed via commit 7e5450eb4a4fa393fb2f537e210b1478acf457cf
